app/lib/backend/storage_rewrite.rb

```diff
@@ -1,106 +1,178 @@
+      def self.rewrite(**kwargs)
+        scope = kwargs[:scope] || self::CLASS
+        ids = kwargs[:ids]
+        scope = scope.where(id: ids) if ids.present?
+        log_progress = kwargs[:log_progress] || false
+        progress = log_progress ? ProgressCounter.new(scope.count) : nil
+
+        scope.includes(self::INCLUDE).find_each do |model|
+          self::REWRITER.call(model)
+          progress&.call
+        end
       end
+    end
 
-      Cinstance.find_each do |cinstance|
+    class CinstanceRewriter < Rewriter
+      CLASS = Cinstance
+      INCLUDE = %i[plan service].freeze
+      REWRITER = ->(cinstance) do
         cinstance.send(:update_backend_application)
         cinstance.send(:update_backend_user_key_to_application_id_mapping)
-
         # to ensure that there is a 'backend_object' - there are
         # invalid data floating around
         if cinstance.provider_account
-          cinstance.application_keys.each { |k| k.send(:update_backend_value) }
-          cinstance.referrer_filters.each { |f| f.send(:update_backend_value) }
+          cinstance.application_keys.each { |app_key| app_key.send(:update_backend_value) }
+          cinstance.referrer_filters.each { |ref_filter| ref_filter.send(:update_backend_value) }
         end
-
-        progress.call
       end
+    end
 
-      Metric.find_each do |metric|
-        metric.sync_backend!
-        progress.call
+    class ServiceRewriter < Rewriter
+      CLASS = Service
+      INCLUDE = :account
+      REWRITER = ->(service) do
+        service.update_backend_service
+        service.service_tokens.find_each(&ServiceTokenService.method(:update_backend))
       end
+    end
 
-      UsageLimit.find_each do |usage_limit|
-        usage_limit.send(:update_backend_usage_limit)
-        progress.call
-      end
+    class MetricRewriter < Rewriter
+      CLASS = Metric
+      INCLUDE = :parent
+      REWRITER = ->(metric) { metric.sync_backend! }
     end
 
-    def rewrite_provider(id)
-      provider = Account.providers_with_master.find(id)
+    class UsageLimitRewriter < Rewriter
+      CLASS = UsageLimit
+      INCLUDE = { plan: :service }.freeze
+      REWRITER = ->(usage_limit) { usage_limit.update_backend_usage_limit }
+    end
+
+    # Used for rewriting all objects inline, and also. `#rewrite` is used by async jobs
+    class Processor
+
+      attr_reader :include_inactive, :log_progress, :action
 
-      services = provider.services.includes(:account)
-      cinstances = provider.buyer_applications.includes(:plan, :service)
+      # @param include_inactive [Boolean] specifies whether to include deleted and suspended providers
+      # @param log_progress [Boolean] specifies whether to print progress to console
+      def initialize(**kwargs)
+        @include_inactive = kwargs[:include_inactive] || false
+        @log_progress = kwargs[:log_progress] || false
+        @action = :rewrite
```

Comment:
this seems to not be used except for logging, and being non-configurable, maybe just hardcode in the log messages?

Reply:
The whole point is that the logs say either "rewrite buyer applications for provider 123..." or "enqueue buyer applications for provider 123...", so you better understand what is being done :upside_down_face:

Reply:
ah, they inherit from each other, I missed that, thanks.